src/Umbraco.Core/Persistence/Repositories/Implement/ContentRepositoryBase.cs

```diff
@@ -835,6 +837,20 @@ protected void PersistRelations(TEntity entity)
                 var val = p.GetValue(); // get the invariant value
                 var refs = reference.GetReferences(val);
                 trackedRelations.AddRange(refs);
+
```

Comment:
This whole section needs to be dealt with differently. I would like to have all of the logic to return `IEnumerable<UmbracoEntityReference>` as tracked relations in a method... probably a method on top of `DataValueReferenceForCollection`. There's a problem with this current logic as you'll see above, it first does this check `if (!(valueEditor is IDataValueReference reference)) continue;` ... and if that doesn't succeed it then moves to the next property which means it entirely ignores any custom `IDataValueReferenceFor` implementations.

So ideally, all of this logic is handled with a call to a new method: `_dataValueReferenceFors.GetAllReferences(entity.Properties);`

Reply:
OK the renames is done to `IDataValueReferenceFactory` and the example code to inherit from both interfaces from your second suggestion.

Why is the method on the Collection as opposed to the Interface/item in the collection @Shazwazza

Reply:
@warrenbuckley I don't really understand your question. What Needs to happen is that this whole highlighted code area: https://github.com/umbraco/Umbraco-CMS/pull/7228/files#diff-83e79f7101cd6797ca42ad8ed07f5835R828-R854 needs to be in a centralized method. A single method used to get all references from all properties for this content item. Not really sure what interface you are referring to here.